File: backend-node/game.js
import { Chess } from "chess.js";

export class Game {
  constructor(player1, player2) {
    this.player1 = player1;
    this.player2 = player2;
    this.board = new Chess();
    this.moves = [];
    this.startTime = new Date();
    this.ended = false;
    this.player1.send(JSON.stringify({ type: "game_start", color: "white" }));
    this.player2.send(JSON.stringify({ type: "game_start", color: "black" }));
    this.rtc = {
      pendingVoiceRequest: false,
      pendingVideoRequest: false,
      voiceAvailable: false,
      videoAvailable: false,
      player1Voice: false,
      player2Voice: false,
      player1Video: false,
      player2Video: false,
    };
  }

  notifyOpponentLeft(disconnectedSocket) {
    if (this.ended) return;
    this.ended = true;
    const opponent =
      disconnectedSocket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(JSON.stringify({ type: "opponent_disconnected" }));
    } catch {
      /* opponent already gone */
    }
  }

  sendChat(socket, message) {
    if (this.ended) return;
    if (!message || message.trim() === "" || message.length > 200) return;

    const from = socket === this.player1 ? "white" : "black";
    const payload = JSON.stringify({
      type: "chat",
      from,
      message: message.trim(),
    });

    this.player1.send(payload);
    this.player2.send(payload);
  }

  makeMove(socket, move) {
    if (this.ended) return;

    if (this.board.turn() === "w" && socket !== this.player1) return;
    if (this.board.turn() === "b" && socket !== this.player2) return;

    try {
      this.board.move(move);
    } catch {
      return;
    }

    this.player1.send(JSON.stringify({ type: "move_made", move }));
    this.player2.send(JSON.stringify({ type: "move_made", move }));

    if (this.board.isGameOver()) {
      this.ended = true;
      let result;

      if (this.board.isCheckmate()) {
        const winner = this.board.turn() === "w" ? "black" : "white";
        result = { winner, reason: "checkmate" };
      } else if (this.board.isStalemate()) {
        result = { winner: "draw", reason: "stalemate" };
      } else if (this.board.isInsufficientMaterial()) {
        result = { winner: "draw", reason: "insufficient_material" };
      } else if (this.board.isThreefoldRepetition()) {
        result = { winner: "draw", reason: "threefold_repetition" };
      } else if (this.board.isDraw()) {
        result = { winner: "draw", reason: "50_move_rule" };
      } else {
        result = { winner: "draw", reason: "unknown" };
      }

      this.player1.send(JSON.stringify({ type: "game_over", result }));
      this.player2.send(JSON.stringify({ type: "game_over", result }));
    }
  }

  handleVoiceRequest(socket) {
    if (this.ended) return;
    this.rtc.pendingVoiceRequest = true;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(JSON.stringify({ type: "voice_request" }));
      console.log("Voice request sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
  }

  handleVoiceResponse(socket, data) {
    if (this.ended || this.rtc.pendingVoiceRequest === false) return;
    console.log("Voice response received:", data);
    if (!data) {
      console.error("Invalid voice response data");
      return;
    }

    this.rtc.voiceAvailable = data.accepted;
    try {
      const opponent = socket === this.player1 ? this.player2 : this.player1;
      opponent.send(
        JSON.stringify({ type: "voice_response", accepted: data.accepted }),
      );
      console.log("Voice response sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      console.log("Failed to send voice response to opponent");
      return;
    }
  }

  handleVideoRequest(socket) {
    if (this.ended) return;
    this.rtc.pendingVideoRequest = true;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(JSON.stringify({ type: "video_request" }));
      console.log("Video request sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      console.log("Failed to send video request to opponent");
      return;
    }
  }

  handleVideoResponse(socket, data) {
    if (this.ended || this.rtc.pendingVideoRequest === false) {
      console.log(
        "Video response received but game ended or no pending request",
      );
      return;
    }
    console.log("Video response received:", data);
    // Expect `{ accepted: true|false }` — accept both answers and declines.
    if (!data) {
      console.error("Invalid video response data");
      // Clear pending flag to avoid leaving the game in a waiting state.
      this.rtc.pendingVideoRequest = false;
      return;
    }

    // Update rtc state and clear pending flag.
    this.rtc.videoAvailable = data.accepted;
    this.rtc.pendingVideoRequest = false;

    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(
        JSON.stringify({ type: "video_response", accepted: data.accepted }),
      );
      console.log("Video response sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
    }
  }

  handleRTCOffer(socket, message) {
    if (this.ended) return;
    if (!this.rtc.voiceAvailable && !this.rtc.videoAvailable) return;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(JSON.stringify({ type: "rtc_offer", sdp: message.sdp }));
      console.log("RTC offer sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
  }

  handleRTCAnswer(socket, message) {
    if (this.ended) return;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(JSON.stringify({ type: "rtc_answer", sdp: message.sdp }));
      console.log("RTC answer sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
  }

  handleRtcIce(socket, message) {
    if (this.ended) return;
    if (!this.rtc.voiceAvailable && !this.rtc.videoAvailable) return;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(
        JSON.stringify({ type: "rtc_ice", candidate: message.candidate }),
      );
      console.log("RTC ice sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
  }

  handleMicToggle(socket, data) {
    if (this.ended) return;
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    try {
      opponent.send(
        JSON.stringify({ type: "mic_toggle", enabled: data.enabled }),
      );
      console.log("Mic toggle sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
    if (this.rtc.voiceAvailable) {
      if (socket === this.player1) {
        this.rtc.player1Voice = data.enabled;
      } else if (socket === this.player2) {
        this.rtc.player2Voice = data.enabled;
      }
    }
  }

  handleCameraToggle(socket, data) {
    if (this.ended) {
      console.log("Game has ended.");
      return;
    }
    const opponent = socket === this.player1 ? this.player2 : this.player1;
    console.log("Camera toggle data:", data);
    try {
      opponent.send(
        JSON.stringify({ type: "camera_toggle", enabled: data.enabled }),
      );
      console.log("Camera toggle sent to opponent");
    } catch {
      console.error("opponent socket is closed or not available");
      return;
    }
    console.log("videoAvailable:", this.rtc.videoAvailable);
    if (this.rtc.videoAvailable) {
      if (socket === this.player1) {
        this.rtc.player1Video = data.enabled;
      } else if (socket === this.player2) {
        this.rtc.player2Video = data.enabled;
      }
    }
  }
  handleResign(socket) {
    if (this.ended) return;
    this.ended = true;
    const winner = socket === this.player1 ? "black" : "white";
    this.player1.send(
      JSON.stringify({
        type: "game_over",
        result: { winner, reason: "resign" },
      }),
    );
    this.player2.send(
      JSON.stringify({
        type: "game_over",
        result: { winner, reason: "resign" },
      }),
    );
  }
}

File: backend-springboot/src/main/java/com/chess/backend/chess/Game.java
package com.chess.backend.chess;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.bhlangonijr.chesslib.Board;
import com.github.bhlangonijr.chesslib.Piece;
import com.github.bhlangonijr.chesslib.Side;
import com.github.bhlangonijr.chesslib.Square;
import com.github.bhlangonijr.chesslib.move.Move;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Port of game.js. Move parsing note: this accepts moves as either a
 * {from, to, promotion} JSON object, or a plain 4-5 character UCI string
 * ("e2e4", "e7e8q") — NOT full SAN ("Nf3"). If your frontend's chess.js
 * integration sends SAN strings, either switch it to send
 * {from, to, promotion} objects (the format the Django/Node versions relied
 * on primarily too), or add a SAN parser here using chesslib's move-generation
 * + comparison utilities.
 */
public class Game {

    private static final Logger log = LoggerFactory.getLogger(Game.class);

    private final ObjectMapper mapper = new ObjectMapper();

    private final WebSocketSession player1; // white
    private final WebSocketSession player2; // black
    private final Board board = new Board();
    private boolean ended = false;

    private boolean pendingVoiceRequest = false;
    private boolean pendingVideoRequest = false;
    private boolean voiceAvailable = false;
    private boolean videoAvailable = false;

    public Game(WebSocketSession player1, WebSocketSession player2) {
        this.player1 = player1;
        this.player2 = player2;
    }

    public WebSocketSession getPlayer1() { return player1; }
    public WebSocketSession getPlayer2() { return player2; }
    public boolean isEnded() { return ended; }

    public void start() {
        send(player1, Map.of("type", "game_start", "color", "white"));
        send(player2, Map.of("type", "game_start", "color", "black"));
    }

    public void notifyOpponentLeft(WebSocketSession disconnected) {
        if (ended) return;
        ended = true;
        WebSocketSession opponent = disconnected.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "opponent_disconnected"));
    }

    public void sendChat(WebSocketSession socket, String message) {
        if (ended) return;
        if (message == null || message.trim().isEmpty() || message.length() > 200) return;

        String from = socket.equals(player1) ? "white" : "black";
        Map<String, Object> payload = Map.of("type", "chat", "from", from, "message", message.trim());
        send(player1, payload);
        send(player2, payload);
    }

    public void makeMove(WebSocketSession socket, JsonNode moveNode) {
        if (ended) return;

        boolean whiteToMove = board.getSideToMove() == Side.WHITE;
        if (whiteToMove && !socket.equals(player1)) return;
        if (!whiteToMove && !socket.equals(player2)) return;

        if (!tryMove(moveNode)) return;

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("type", "move_made");
        payload.put("move", jsonToObject(moveNode));
        send(player1, payload);
        send(player2, payload);

        if (isGameOver()) {
            ended = true;
            Map<String, Object> result = determineResult();
            Map<String, Object> overPayload = Map.of("type", "game_over", "result", result);
            send(player1, overPayload);
            send(player2, overPayload);
        }
    }

    private boolean tryMove(JsonNode moveNode) {
    try {
        if (moveNode.isTextual()) {
            // chesslib's Board.doMove(String) parses standard SAN
            // directly (e.g. "e4", "Nf3", "O-O") — no manual parsing needed.
            return board.doMove(moveNode.asText().trim());
        }

        if (moveNode.isObject()) {
            String from = moveNode.path("from").asText(null);
            String to = moveNode.path("to").asText(null);
            String promotion = moveNode.path("promotion").asText(null);
            if (from == null || to == null) return false;

            Square fromSq = Square.fromValue(from.toUpperCase());
            Square toSq = Square.fromValue(to.toUpperCase());
            Piece promoPiece = Piece.NONE;
            if (promotion != null && !promotion.isBlank()) {
                promoPiece = resolvePromotionPiece(promotion, board.getSideToMove());
            }

            Move move = new Move(fromSq, toSq, promoPiece);
            return board.doMove(move, true);
        }

        return false;
    } catch (Exception e) {
        return false;
    }
}

    private Piece resolvePromotionPiece(String code, Side side) {
        boolean white = side == Side.WHITE;
        return switch (code.toLowerCase()) {
            case "q" -> white ? Piece.WHITE_QUEEN : Piece.BLACK_QUEEN;
            case "r" -> white ? Piece.WHITE_ROOK : Piece.BLACK_ROOK;
            case "b" -> white ? Piece.WHITE_BISHOP : Piece.BLACK_BISHOP;
            case "n" -> white ? Piece.WHITE_KNIGHT : Piece.BLACK_KNIGHT;
            default -> Piece.NONE;
        };
    }

    private boolean isGameOver() {
        return board.isMated()
                || board.isStaleMate()
                || board.isInsufficientMaterial()
                || board.isRepetition()
                || board.getHalfMoveCounter() >= 100;
    }

    private Map<String, Object> determineResult() {
        Map<String, Object> result = new LinkedHashMap<>();
        if (board.isMated()) {
            String winner = board.getSideToMove() == Side.WHITE ? "black" : "white";
            result.put("winner", winner);
            result.put("reason", "checkmate");
        } else if (board.isStaleMate()) {
            result.put("winner", "draw");
            result.put("reason", "stalemate");
        } else if (board.isInsufficientMaterial()) {
            result.put("winner", "draw");
            result.put("reason", "insufficient_material");
        } else if (board.isRepetition()) {
            result.put("winner", "draw");
            result.put("reason", "threefold_repetition");
        } else if (board.getHalfMoveCounter() >= 100) {
            result.put("winner", "draw");
            result.put("reason", "50_move_rule");
        } else {
            result.put("winner", "draw");
            result.put("reason", "unknown");
        }
        return result;
    }

    public void handleVoiceRequest(WebSocketSession socket) {
        if (ended) return;
        pendingVoiceRequest = true;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "voice_request"));
        log.info("Voice request sent to opponent");
    }

    public void handleVoiceResponse(WebSocketSession socket, JsonNode data) {
        if (ended || !pendingVoiceRequest) return;
        log.info("Voice response received: {}", data);
        if (data == null) {
            log.error("Invalid voice response data");
            return;
        }
        boolean accepted = data.path("accepted").asBoolean(false);
        voiceAvailable = accepted;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "voice_response", "accepted", accepted));
        log.info("Voice response sent to opponent");
    }

    public void handleVideoRequest(WebSocketSession socket) {
        if (ended) return;
        pendingVideoRequest = true;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "video_request"));
        log.info("Video request sent to opponent");
    }

    public void handleVideoResponse(WebSocketSession socket, JsonNode data) {
        if (ended || !pendingVideoRequest) {
            log.info("Video response received but game ended or no pending request");
            return;
        }
        log.info("Video response received: {}", data);
        if (data == null) {
            log.error("Invalid video response data");
            pendingVideoRequest = false;
            return;
        }
        boolean accepted = data.path("accepted").asBoolean(false);
        videoAvailable = accepted;
        pendingVideoRequest = false;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "video_response", "accepted", accepted));
        log.info("Video response sent to opponent");
    }

    public void handleRtcOffer(WebSocketSession socket, JsonNode message) {
        if (ended) return;
        if (!voiceAvailable && !videoAvailable) return;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "rtc_offer", "sdp", message.path("sdp").asText(null)));
        log.info("RTC offer sent to opponent");
    }

    public void handleRtcAnswer(WebSocketSession socket, JsonNode message) {
        if (ended) return;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "rtc_answer", "sdp", message.path("sdp").asText(null)));
        log.info("RTC answer sent to opponent");
    }

    public void handleRtcIce(WebSocketSession socket, JsonNode message) {
        if (ended) return;
        if (!voiceAvailable && !videoAvailable) return;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "rtc_ice", "candidate", jsonToObject(message.path("candidate"))));
        log.info("RTC ice sent to opponent");
    }

    public void handleMicToggle(WebSocketSession socket, JsonNode data) {
        if (ended) return;
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        send(opponent, Map.of("type", "mic_toggle", "enabled", data.path("enabled").asBoolean(false)));
        log.info("Mic toggle sent to opponent");
    }

    public void handleCameraToggle(WebSocketSession socket, JsonNode data) {
        if (ended) {
            log.info("Game has ended.");
            return;
        }
        WebSocketSession opponent = socket.equals(player1) ? player2 : player1;
        log.info("Camera toggle data: {}", data);
        send(opponent, Map.of("type", "camera_toggle", "enabled", data.path("enabled").asBoolean(false)));
        log.info("Camera toggle sent to opponent");
    }

    public void handleResign(WebSocketSession socket) {
        if (ended) return;
        ended = true;
        String winner = socket.equals(player1) ? "black" : "white";
        Map<String, Object> payload = Map.of(
                "type", "game_over",
                "result", Map.of("winner", winner, "reason", "resign"));
        send(player1, payload);
        send(player2, payload);
    }

    // ── helpers ──

    private void send(WebSocketSession session, Map<String, Object> payload) {
        try {
            if (session != null && session.isOpen()) {
                session.sendMessage(new TextMessage(mapper.writeValueAsString(payload)));
            }
        } catch (Exception e) {
            log.error("opponent socket is closed or not available");
        }
    }

    private Object jsonToObject(JsonNode node) {
        try {
            return mapper.treeToValue(node, Object.class);
        } catch (Exception e) {
            return null;
        }
    }
}
